lambda/subarea/POST/index.js

```diff
@@ -20,40 +37,112 @@ exports.handler = async (event, context) => {
       return sendResponse(403, { msg: 'Unauthorized.' }, context);
     }
 
-    if (event?.queryStringParameters?.type === "config") {
-      return await handleConfig(body, context);
-    } else if (event?.queryStringParameters?.type === "activity") {
-      // Handle standard monthly entry for this orc::subAreaName::activity
-      return await handleActivity(body);
-    } else {
-      throw "Invalid request";
+    // check if fiscal year is locked
+    if (await checkFiscalYearLock(body)) {
+      return sendResponse(403, { msg: `This fiscal year has been locked against editing by the system administrator.` }, context)
+    }
+
+    // Disabling the ability to change config for now. 
+    // Request used to necessitate 'type = activity/config' as a queryParam (was future proofing).
+    // Queryparams no longer required. All info included in request body. 
+    // Refer to code prior to 2022-09-27 for handleConfig.
+
+    await verifyBody(body);
+
+    // check if record is locked
+    const unlocking = lock === false ? true : false;
+    const existingRecord = await getOne(`${body.subAreaId}::${body.activity}`, body.date);
+    if (existingRecord?.isLocked && !unlocking) {
+      return sendResponse(403, { msg: 'Record is locked.' });
     }
+
+    // handle locking/unlocking existing records
+    if (lock !== null) {
+      if (existingRecord?.pk) {
+        return await handleLockUnlock(existingRecord, lock, context);
+      } else if (lock === false) {
+        // if record doesnt exist, we can't unlock it
+        return sendResponse(404, { msg: 'Record not found.' });
+      }
+      // if we are locking a record that doesn't exist, we need to create it.
+      // fall through and create new record for locking.
+    }
+    return await handleActivity(body, context);
   } catch (err) {
     logger.error(err);
     return sendResponse(400, { msg: "Invalid request" }, context);
   }
 };
 
+async function checkFiscalYearLock(body) {
+  // extract fiscal year from date
+  let recordYear = Number(body.date.slice(0, 4));
+  let recordMonth = Number(body.date.slice(4, 6));
+  if (recordMonth > FISCAL_YEAR_FINAL_MONTH) {
+    recordYear++;
+  }
+  const fiscalYearEndObj = await getOne('fiscalYearEnd', String(recordYear));
+  if (fiscalYearEndObj?.isLocked) {
+    return true;
+  }
+  return false;
+}
+
+async function verifyBody(body) {
+  if (!body.subAreaId || !body.activity || !body.date) {
+    return sendResponse(400, { msg: "Invalid request." });
+  };
+
+  // delete isLocked - need correct path to lock/unlock records
+  delete body.isLocked;
+}
+
+async function handleLockUnlock(record, lock, context) {
+  const updateObj = {
+    TableName: TABLE_NAME,
+    Key: {
+      pk: { S: record.pk },
+      sk: { S: record.sk }
+    },
+    UpdateExpression: 'set isLocked = :isLocked',
+    ExpressionAttributeValues: {
+      ':isLocked': { BOOL: lock }
+    },
+    ReturnValues: 'ALL_NEW'
+  };
```

Comment:
We may want to put a `ConditionExpression` of `isLocked = !lock`.  That way when you fail you can return `HTTP 409`.

Reply:
Fixed in `89812d6`.